src/Fl_SVG_Image.cxx

```diff
@@ -90,143 +144,165 @@ float Fl_SVG_Image::svg_scaling_(int W, int H) {
-      p += l; *p = 0;
-    }
-  } while (is_compressed && l >0);
-  gzclose(gzf);
-  if (is_compressed) out = (char*)realloc(out, (p-out)+1);
-  return out;
+  // delete all the chunks that we allocated
+  chunk = first;
+  while (chunk) {
+    next_chunk = chunk->next;
+    delete chunk;
+    chunk = next_chunk;
+  }
+
+  return err == Z_STREAM_END ? Z_OK :
+  err == Z_NEED_DICT ? Z_DATA_ERROR  :
+  err == Z_BUF_ERROR && stream.avail_out ? Z_DATA_ERROR :
+  err;
 }
 
+
 #endif // defined(HAVE_LIBZ)
 
 
-void Fl_SVG_Image::init_(const char *filename, const unsigned char *in_filedata, const Fl_SVG_Image *copy_source, size_t length) {
-  if (copy_source) {
-    filename = NULL;
-    in_filedata = NULL;
-    counted_svg_image_ = copy_source->counted_svg_image_;
-    counted_svg_image_->ref_count++;
-  } else {
-    counted_svg_image_ = new counted_NSVGimage;
-    counted_svg_image_->svg_image = NULL;
-    counted_svg_image_->ref_count = 1;
-  }
-  char *filedata = NULL;
+void Fl_SVG_Image::init_(const char *name, const unsigned char *in_data, size_t length) {
+  counted_svg_image_ = new counted_NSVGimage;
+  counted_svg_image_->svg_image = NULL;
+  counted_svg_image_->ref_count = 1;
   to_desaturate_ = false;
   average_weight_ = 1;
   proportional = true;
-  bool is_compressed = true;
 
-  if (filename || length) { // process file or byte buffer
-#if defined(HAVE_LIBZ)
-    int fdread, fdwrite = -1;
-    if (length) { // process gzip'ed byte buffer
-      // Pipe gzip'ed byte buffer into gzlib inflate algorithm.
-      // Under Windows, gzip'ed byte buffer is written to pipe by child thread.
-      // Under Posix, gzip'ed byte buffer is written to pipe by non-blocking write
-      // done by main thread.
-      Fl::system_driver()->pipe_support(fdread, fdwrite, in_filedata, length);
-    } else { // read or decompress a .svg or .svgz file
-      struct stat svg_file_stat;
-      fl_stat(filename, &svg_file_stat); // get file size
-      fdread = fl_open_ext(filename, 1, 0);
-      // read a possibly gzip'ed file and return result as char string
-      length = svg_file_stat.st_size;
-      is_compressed = (strcmp(filename + strlen(filename) - 5, ".svgz") == 0);
-    }
-    gzFile gzf =  (fdread >= 0 ? gzdopen(fdread, "rb") : NULL);
-    if (gzf) {
-      filedata = svg_inflate(gzf, length, is_compressed, fdwrite, in_filedata);
-    } else {
-      if (fdread >= 0) Fl::system_driver()->close_fd(fdread);
-      if (fdwrite >= 0) Fl::system_driver()->close_fd(fdwrite);
-    }
-
-#else // ! HAVE_LIBZ
-    // without libz, read .svg file
-    FILE *fp = fl_fopen(filename, "rb");
-    if (fp) {
-      fseek(fp, 0, SEEK_END);
-      long size = ftell(fp);
-      fseek(fp, 0, SEEK_SET);
-      filedata = (char*)malloc(size+1);
-      if (filedata) {
-        if (fread(filedata, 1, size, fp) == size) {
-          filedata[size] = '\0';
+  // yes, this is a const cast to avoid duplicating user supplied data
+  uchar *data = (uchar*)in_data;
```

Comment:
Then we should IMHO use `const_cast` explicitly in new code, according to recent discussion in fltk.general and the new CMP (will be updated very soon).

Reply:
For good measure, I added an emoji too, since we do that now.